frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -751,6 +750,44 @@ pub mod pallet {
             let staker_reward =
                 Perbill::from_rational(staked, staking_info.total) * stakers_joint_reward;
 
+            let mut ledger = Self::ledger(&staker);
+
+            if Self::should_restake_reward(
+                ledger.reward_destination,
+                &contract_id,
+                staker_info.latest_staked_value(),
+            ) {
+                // There must be one slot left for restaking
+                ensure!(
+                    staker_info.len() < T::MaxEraStakeValues::get(),
+                    Error::<T>::TooManyEraStakeValues
+                );
```

Comment:
in case of a batch call, only the last item would be mutated, right?
so, something like:
```rust
ensure!(
    staker_info.len() < T::MaxEraStakeValues::get() 
    || staker_info.latest_staked_era == current_era,
    Error::<T>::TooManyEraStakeValues
);
```
should make the result:
```
1. [<1000, 6>, <2000, 10>, <2001, 15>] 
2. [<1000, 7>, <2000, 10>, <2002, 15>] 
3. [<1000, 8>, <2000, 10>, <2003, 15>]
4. [<1000, 9>, <2000, 10>, <2004, 15>]
5. [<2000, 10>, <2005, 15>]
6. [<2000, 11>, <2007, 15>]
7. [<2000, 12>, <2009, 15>]
8. [<2000, 13>, <2011, 15>]
9.[<2000,14>,<2013,15>]
10.[<2015, 15>]
```

Reply:
Yes, that's correct.
I gave this batch example just for the demonstration of how it should work.
You don't need to use `last_staked_era` for this check. :)

Just use the exact same check as in `bond_and_stake` and `unbond_and_unstake`.
With a small difference: use `<=` instead of `<`.

Just move L761 code below L766.

Reply:
I extracted the condition and moved it before the call to `T::Currency::withdraw` because the body of the `if` clause also had possible errors.